264441: incorrect hostname in SASL bind within rebind function

Source: upstream, http://www.openldap.org/its/index.cgi/Software%20Bugs?id=3298
Reviewed-By: Jan Safranek <jsafrane@redhat.com>

ldap_int_sasl_bind() used wrong hostname when chasing referal.

--- openldap-2.2.13/libraries/libldap/cyrus.c	2004-05-21 12:29:34.000000000 -0500
+++ openldap-2.2.13/libraries/libldap/cyrus.c.new	2007-08-29 11:58:22.000000000 -0500
@@ -593,7 +593,7 @@
 		rc = ldap_open_defconn( ld );
 		if( rc < 0 ) return ld->ld_errno;
 
-		ber_sockbuf_ctrl( ld->ld_sb, LBER_SB_OPT_GET_FD, &sd );
+		ber_sockbuf_ctrl( ld->ld_defconn->lconn_sb, LBER_SB_OPT_GET_FD, &sd );
 
 		if( sd == AC_SOCKET_INVALID ) {
 			ld->ld_errno = LDAP_LOCAL_ERROR;
@@ -611,7 +611,7 @@
 		ld->ld_defconn->lconn_sasl_authctx = NULL;
 	}
 
-	{ char *saslhost = ldap_host_connected_to( ld->ld_sb, "localhost" );
+	{ char *saslhost = ldap_host_connected_to( ld->ld_defconn->lconn_sb, "localhost" );
 	rc = ldap_int_sasl_open( ld, ld->ld_defconn, saslhost );
 	LDAP_FREE( saslhost );
 	}
@@ -621,7 +621,7 @@
 	ctx = ld->ld_defconn->lconn_sasl_authctx;
 
 	/* Check for TLS */
-	ssl = ldap_pvt_tls_sb_ctx( ld->ld_sb );
+	ssl = ldap_pvt_tls_sb_ctx( ld->ld_defconn->lconn_sb );
 	if ( ssl ) {
 		struct berval authid = BER_BVNULL;
 		ber_len_t fac;
@@ -821,7 +821,7 @@
 			if ( ld->ld_defconn->lconn_sasl_sockctx ) {
 				oldctx = ld->ld_defconn->lconn_sasl_sockctx;
 				sasl_dispose( &oldctx );
-				ldap_pvt_sasl_remove( ld->ld_sb );
+				ldap_pvt_sasl_remove( ld->ld_defconn->lconn_sb );
 			}
 			ldap_pvt_sasl_install( ld->ld_conns->lconn_sb, ctx );
 			ld->ld_defconn->lconn_sasl_sockctx = ctx;
